### PR TITLE
Beacon node slasher client infrastructure

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #3133

adding slasher client into beacon node in order to request slashing events to include in blocks.
this is just the infrastructure for the features to come

[slasher beacon node communication design diagram](https://docs.google.com/document/d/1DWDvht3-FmvvDNskzmgN26_TZIs35lo66e53XqdViy8/edit?usp=sharing)

### CODE REVIEW

File: beacon-chain/blockchain/forkchoice/process_attestation.go [hunk 1]
```diff
@@ -166,12 +166,29 @@ func (s *Store) verifyAttestation(ctx context.Context, baseState *pb.BeaconState
 	if err != nil {
 		return nil, errors.Wrap(err, "could not convert attestation to indexed attestation")
 	}
+	// check slashing
+	go s.checkForSlashing(ctx, indexedAtt)
```
Comment (terencechain) [inline beacon-chain/blockchain/forkchoice/process_attestation.go]: I would avoid doing this. We should feed the attestation from sync service to slasher service and run it there. It's important to keep the state transition pure and well contained. The cost of converting attestation to indexed attestation is almost non-negligible so it really makes no sense to plug slasher here

File: beacon-chain/blockchain/forkchoice/service.go [hunk 2]
```diff
@@ -41,18 +41,20 @@ type Store struct {
...
-func NewForkChoiceService(ctx context.Context, db db.Database) *Store {
+func NewForkChoiceService(ctx context.Context, db db.Database, sc ethpb.SlasherClient) *Store {
 	ctx, cancel := context.WithCancel(ctx)
 	return &Store{
 		ctx:             ctx,
 		cancel:          cancel,
 		db:              db,
 		checkpointState: cache.NewCheckpointStateCache(),
 		seenAtts:        make(map[[32]byte]bool),
+		slasherClient:   sc,
```
Comment (terencechain) [inline beacon-chain/blockchain/forkchoice/service.go]: Same here as feeding slasher client to fork choice service. It feels like a missuse of responsibility

File: beacon-chain/blockchain/service.go [hunk 3]
```diff
@@ -71,13 +71,14 @@ type Config struct {
 	P2p               p2p.Broadcaster
 	MaxRoutines       int64
 	StateNotifier     statefeed.Notifier
+	SlasherClient     ethpb.SlasherClient
```
Comment (terencechain) [inline beacon-chain/blockchain/service.go]: Same as the comment above

**Comment by rauljordan** [comment]
Is this still relevant with #4230 @shayzluf ?

**Comment by shayzluf** [comment]
@rauljordan this is the other way around. i will clean it up and make it ready again

File: shared/featureconfig/flags.go [hunk 4]
```diff
@@ -85,6 +85,11 @@ var (
 			"initial sync and disk-IO is one of the biggest bottleneck. This still saves finalized state in DB " +
 			"and start syncing from there",
 	}
+	connectToSlasherFlag = cli.BoolFlag{
```
Comment (terencechain) [inline shared/featureconfig/flags.go]: Please provide more context, ie. how is the slasher connected with the beacon node?

File: shared/featureconfig/flags.go [hunk 5]
```diff
@@ -179,4 +184,5 @@ var BeaconChainFlags = append(deprecatedFlags, []cli.Flag{
 	enableSnappyDBCompressionFlag,
 	enablePruneBoundaryStateFlag,
 	fastCommitteeAssignmentsFlag,
+	connectToSlasherFlag,
```
Comment (terencechain) [inline shared/featureconfig/flags.go]: This name can be better, enableSlasherFlag?

File: beacon-chain/rpc/service.go [hunk 6]
```diff
@@ -252,6 +264,44 @@ func (s *Service) Start() {
...
+			grpc_opentracing.UnaryClientInterceptor(),
+			grpc_prometheus.UnaryClientInterceptor,
+		)),
+	}
+	conn, err := grpc.DialContext(s.ctx, s.slasherProvider, slasherOpts...)
+	if err != nil {
+		log.Errorf("Could not dial endpoint: %s, %v", s.slasherProvider, err)
+		return
+	}
+	log.Info("Successfully started gRPC connection")
```
Comment (terencechain) [inline beacon-chain/rpc/service.go]: This log is too vague, gRPC connection to where?

File: beacon-chain/rpc/service.go [hunk 7]
```diff
@@ -261,13 +311,23 @@ func (s *Service) Stop() error {
...
+	}
 	return nil
 }
 
 // Status returns nil or credentialError
 func (s *Service) Status() error {
+	var err error
 	if s.credentialError != nil {
-		return s.credentialError
+		err = s.credentialError
```
Comment (terencechain) [inline beacon-chain/rpc/service.go]: I dont understand why not return the error here?

File: beacon-chain/rpc/service.go [hunk 8]
```diff
@@ -8,10 +8,13 @@ import (
 	"net"
 	"os"
 
+	"github.com/prysmaticlabs/prysm/shared/featureconfig"
+
```
Comment (terencechain) [inline beacon-chain/rpc/service.go]: go imports

File: beacon-chain/flags/base.go [hunk 9]
```diff
@@ -56,4 +56,15 @@ var (
...
+	// SlasherCertFlag defines a flag for the slasher TLS certificate.
+	SlasherCertFlag = cli.StringFlag{
+		Name:  "slasher-tls-cert",
+		Usage: "Certificate for secure slasher gRPC connection. Pass this in order to use slasher gRPC securely.",
+	}
+	// SlasherProviderFlag defines a flag for a slasher RPC provider.
+	SlasherProviderFlag = cli.StringFlag{
+		Name:  "slasher-provider",
+		Usage: "A slasher provider string endpoint. Can either be an grpc server endpoint.",
+		Value: "127.0.0.1:5000",
```
Comment (terencechain) [inline beacon-chain/flags/base.go]: why 5000?

File: .gitignore [hunk 10]
```diff
@@ -26,3 +26,6 @@ password.txt
 # go dependancy
 /go.mod
 /go.sum
+
+# Slasher tmp db
+/slasherdata/slasher.db
```
Comment (terencechain) [inline .gitignore]: Why do we have to add this under gitIgnore? We dont do this for beacon DB

**Comment by shayzluf** [comment]
> Just a few questions.
> 
> Do you have a design doc? feel free to link that in the descriptions

thanks added communication diagram to pr description

File: shared/featureconfig/flags.go (hunk 5 above)
Comment (shayzluf) [reply shared/featureconfig/flags.go]: changed

File: shared/featureconfig/flags.go (hunk 4 above)
Comment (shayzluf) [reply shared/featureconfig/flags.go]: added context

File: beacon-chain/rpc/service.go (hunk 6 above)
Comment (shayzluf) [reply beacon-chain/rpc/service.go]: updated

File: beacon-chain/rpc/service.go (hunk 7 above)
Comment (shayzluf) [reply beacon-chain/rpc/service.go]: if there are two credentials errors the user can resolve both of them

Comment (shayzluf) [reply beacon-chain/rpc/service.go]: changed back

File: beacon-chain/rpc/service.go (hunk 8 above)
Comment (shayzluf) [reply beacon-chain/rpc/service.go]: done. 🙏

File: beacon-chain/flags/base.go (hunk 9 above)
Comment (shayzluf) [reply beacon-chain/flags/base.go]: i don't know . just a number other than 4000

File: .gitignore (hunk 10 above)
Comment (shayzluf) [reply .gitignore]: removed

**Pull request merged**